File: FarNet/Viewer.cs
/*
FarNet plugin for Far Manager
Copyright (c) 2005-2009 FarNet Team
*/

using System.Diagnostics;
using System;

namespace FarNet
{
	/// <summary>
	/// Viewer base interface. Exposed as <see cref="IFar.AnyViewer"/>.
	/// </summary>
	public interface IAnyViewer
	{
		/// <summary>
		/// A file is opened in the viewer.
		/// </summary>
		/// <remarks>
		/// This event can be triggered more than once for the same viewer instance,
		/// e.g. by 'Add', 'Subtract' keys.
		/// </remarks>
		event EventHandler Opened;
		/// <summary>
		/// Viewer is closed. Don't operate on it, it has really gone.
		/// </summary>
		/// <remarks>
		/// This event is triggered once for the viewer instance,
		/// even if there were several files opened in it.
		/// </remarks>
		event EventHandler Closed;
		/// <summary>
		/// Viewer window has got focus. FAR 1.71.2406
		/// </summary>
		event EventHandler GotFocus;
		/// <summary>
		/// Viewer window is losing focus. FAR 1.71.2406
		/// </summary>
		event EventHandler LosingFocus;
		/// <summary>
		/// Opens a viewer to view some text.
		/// </summary>
		void ViewText(string text, string title, OpenMode mode);
	}

	/// <summary>
	/// Viewer interface. Exposed as <see cref="IFar.Viewer"/>. Created by <see cref="IFar.CreateViewer"/>.
	/// </summary>
	/// <remarks>
	/// Normally this object should be created or requested, used instantly and never kept for future use.
	/// When you need the current viewer instance next time call <see cref="IFar.Viewer"/> again to get it.
	/// </remarks>
	public interface IViewer : IAnyViewer
	{
		/// <summary>
		/// Internal ID.
		/// </summary>
		int Id { get; }
		/// <summary>
		/// Option to delete a source file when the viewer is closed.
		/// </summary>
		DeleteSource DeleteSource { get; set; }
		/// <summary>
		/// Switching between editor and viewer. Set it before opening.
		/// </summary>
		Switching Switching { get; set; }
		/// <summary>
		/// Do not use viewer history. Set it before opening.
		/// </summary>
		bool DisableHistory { get; set; }
		/// <summary>
		/// Name of a file being viewed. Set it before opening.
		/// On opening it can be corrected, e.g. converted into full path.
		/// </summary>
		string FileName { get; set; }
		/// <summary>
		/// Code page identifier. Set it before opening.
		/// </summary>
		int CodePage { get; set; }
		/// <summary>
		/// Window start position. Set it before opening.
		/// </summary>
		Place Window { get; set; }
		/// <summary>
		/// Current viewer window size.
		/// </summary>
		Point WindowSize { get; }
		/// <summary>
		/// Window title. Set it before opening.
		/// </summary>
		string Title { get; set; }
		/// <summary>
		/// Opens the viewer using properties:
		/// <see cref="DeleteSource"/>
		/// <see cref="DisableHistory"/>
		/// <see cref="Switching"/>
		/// <see cref="FileName"/>
		/// <see cref="Title"/>
		/// <see cref="Window"/>
		/// </summary>
		void Open(OpenMode mode);
		/// <summary>
		/// See <see cref="Open(OpenMode)"/> with <see cref="OpenMode.None"/>.
		/// </summary>
		void Open();
		/// <summary>
		/// File size.
		/// </summary>
		long FileSize { get; }
		/// <summary>
		/// View frame.
		/// </summary>
		ViewFrame Frame { get; set; }
		/// <summary>
		/// Sets new viewer frame.
		/// </summary>
		/// <param name="pos">New file position (depends on options).</param>
		/// <param name="left">New left position.</param>
		/// <param name="options">Options.</param>
		/// <returns>New actual position.</returns>
		long SetFrame(long pos, int left, ViewFrameOptions options);
		/// <summary>
		/// Closes the current viewer window.
		/// </summary>
		void Close();
		/// <summary>
		/// Redraws the current viewer window.
		/// </summary>
		void Redraw();
		/// <summary>
		/// Sets selected block.
		/// </summary>
		/// <param name="symbolStart">Selection start in charactes, not in bytes.</param>
		/// <param name="symbolCount">Selected character count.</param>
		void Select(long symbolStart, int symbolCount);
		/// <summary>
		/// Hexadecimal mode.
		/// </summary>
		bool HexMode { get; set; }
		/// <summary>
		/// Wrap mode. 
		/// </summary>
		bool WrapMode { get; set; }
		/// <summary>
		/// Word wrap mode. 
		/// </summary>
		bool WordWrapMode { get; set; }
	}

	/// <summary>
	/// Options for <see cref="IViewer.SetFrame"/>.
	/// </summary>
	[Flags]
	public enum ViewFrameOptions
	{
		///
		None,
		/// <summary>
		/// Don't redraw.
		/// </summary>
		NoRedraw = 1,
		/// <summary>
		/// Position is defined in percents.
		/// </summary>
		Percent = 2,
		/// <summary>
		/// Position is relative to the current (and can be negative).
		/// </summary>
		Relative = 4
	}

	/// <summary>
	/// Open modes of editor and viewer.
	/// </summary>
	public enum OpenMode
	{
		/// <summary>
		/// Tells to open not modal editor or viewer and return immediately.
		/// </summary>
		None,
		/// <summary>
		/// Tells to open not modal editor or viewer and wait for exit.
		/// </summary>
		Wait,
		/// <summary>
		/// Tells to open modal editor or viewer.
		/// </summary>
		Modal
	}

	/// <summary>
	/// Options to delete a file when a viewer or editor is closed.
	/// </summary>
	public enum DeleteSource
	{
		/// <summary>
		/// Default action: do not delete a file.
		/// </summary>
		None,
		/// <summary>
		/// Try to delete a file always. It is not recommended if editor\viewer switching is enabled (F6).
		/// You may set it at any time, i.e. before or after opening.
		/// </summary>
		File,
		/// <summary>
		/// The same as <see cref="File"/> plus delete its folder if it is empty.
		/// You may set it at any time, i.e. before or after opening.
		/// </summary>
		Folder,
		/// <summary>
		/// Delete a file if it was not used. The file is "used" if:
		/// *) it was saved;
		/// *) there was editor\viewer switching (F6).
		/// *) it is opened in another editor or viewer.
		/// You should set it before opening.
		/// </summary>
		UnusedFile,
		/// <summary>
		/// The same as <see cref="UnusedFile"/> plus delete its folder if it is empty.
		/// You should set it before opening.
		/// </summary>
		UnusedFolder
	}

	/// <summary>
	/// Viewer frame info.
	/// </summary>
	[DebuggerStepThroughAttribute]
	public struct ViewFrame
	{
		///
		public ViewFrame(long pos, int left)
		{
			_pos = pos;
			_leftPos = left;
		}
		/// <summary>
		/// Position in the file.
		/// </summary>
		public long Pos { get { return _pos; } set { _pos = value; } }
		long _pos;
		/// <summary>
		/// Leftmost visible position of the text on the screen.
		/// </summary>
		public int LeftPos { get { return _leftPos; } set { _leftPos = value; } }
		int _leftPos;
		///
		public static bool operator ==(ViewFrame left, ViewFrame right)
		{
			return
				left._pos == right._pos &&
				left._leftPos == right._leftPos;
		}
		///
		public static bool operator !=(ViewFrame left, ViewFrame right)
		{
			return !(left == right);
		}
		///
		public override bool Equals(Object obj)
		{
			if (obj == null || GetType() != obj.GetType())
				return false;
			ViewFrame that = (ViewFrame)obj;
			return this == that;
		}
		///
		public override string ToString()
		{
			return "(" + _pos + ", " + _leftPos + ")";
		}
		///
		public override int GetHashCode()
		{
			return base.GetHashCode();
		}
	}
}

File: trunk/FarNet/FarNetMan/Viewer.cpp
/*
FarNet plugin for Far Manager
Copyright (c) 2005-2009 FarNet Team
*/

#include "StdAfx.h"
#include "Viewer.h"
#include "Far.h"
#include "Utils.h"
#include "Viewer0.h"

namespace FarNet
{;
Viewer::Viewer()
: _id(-1)
, _Title(String::Empty)
, _CodePage(CP_AUTODETECT)
{}

void Viewer::Open()
{
	Open(OpenMode::None);
}

void Viewer::Open(OpenMode mode)
{
	AssertClosed();

	// flags
	int flags = 0;
	
	if (_DisableHistory)
		flags |= VF_DISABLEHISTORY;
	
	switch(_Switching)
	{
	case FarNet::Switching::Enabled:
		flags |= VF_ENABLE_F6;
		break;
	case FarNet::Switching::Auto:
		if (_DeleteSource == FarNet::DeleteSource::None &&
			_Closed == nullptr &&
			_GotFocus == nullptr &&
			_LosingFocus == nullptr &&
			_Opened == nullptr)
			flags |= VF_ENABLE_F6;
		break;
	}
	
	switch(mode)
	{
	case OpenMode::None:
		flags |= (VF_NONMODAL | VF_IMMEDIATERETURN); break;
	case OpenMode::Wait:
		flags |= VF_NONMODAL; break;
	}
	
	switch(_DeleteSource)
	{
	case FarNet::DeleteSource::UnusedFile:
		flags |= VF_DELETEONLYFILEONCLOSE; break;
	case FarNet::DeleteSource::UnusedFolder:
		flags |= VF_DELETEONCLOSE; break;
	}

	PIN_ES(pinFileName, _FileName);
	PIN_ES(pinTitle, _Title);

	// from dialog? set modal
	WindowType wt = Far::Instance->WindowType;
	if (wt == WindowType::Dialog)
		flags &= ~VF_NONMODAL;

	// open: see editor
	_id = -1;
	Viewer0::_viewerWaiting = this;
	Info.Viewer(
		pinFileName,
		pinTitle,
		_Window.Left,
		_Window.Top,
		_Window.Right,
		_Window.Bottom,
		flags,
		_CodePage); //?? test window values

	// redraw Far
	if (wt == WindowType::Dialog)
		Far::Instance->Redraw();

	// errors: see editor
	if (_id == -1)
		throw gcnew OperationCanceledException("Cannot open the file '" + FileName + "'");
}

int Viewer::Id::get()
{
	return _id;
}

DeleteSource Viewer::DeleteSource::get()
{
	return _DeleteSource;
}

void Viewer::DeleteSource::set(FarNet::DeleteSource value)
{
	_DeleteSource = value;
}

Switching Viewer::Switching::get()
{
	return _Switching;
}

void Viewer::Switching::set(FarNet::Switching value)
{
	AssertClosed();
	_Switching = value;
}

bool Viewer::DisableHistory::get()
{
	return _DisableHistory;
}

void Viewer::DisableHistory::set(bool value)
{
	AssertClosed();
	_DisableHistory = value;
}

bool Viewer::IsOpened::get()
{
	return _id >= 0;
}

String^ Viewer::FileName::get()
{
	return _FileName;
}

void Viewer::FileName::set(String^ value)
{
	AssertClosed();
	_FileName = value;
}

String^ Viewer::Title::get()
{
	return _Title;
}

Int64 Viewer::FileSize::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	if (vi.ViewerID >= 0 && vi.ViewerID == _id)
		return vi.FileSize;
	else
		return -1;
}

Point Viewer::WindowSize::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	Point r;
	if (vi.ViewerID >= 0 && vi.ViewerID == _id)
	{
		r.X = vi.WindowSizeX;
		r.Y = vi.WindowSizeY;
	}
	return r;
}

void Viewer::Title::set(String^ value)
{
	AssertClosed();
	_Title = value;
}

Place Viewer::Window::get()
{
	return _Window;
}

void Viewer::Window::set(Place value)
{
	AssertClosed();
	_Window = value;
}

ViewFrame Viewer::Frame::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	ViewFrame r;
	if (vi.ViewerID >= 0 && vi.ViewerID == _id)
	{
		r.Pos = vi.FilePos;
		r.LeftPos = vi.LeftPos;
	}
	return r;
}

void Viewer::Frame::set(ViewFrame value)
{
	AssertCurrentViewer();
	ViewerSetPosition vsp;
	vsp.Flags = VSP_NORETNEWPOS;
	vsp.LeftPos = value.LeftPos;
	vsp.StartPos = value.Pos;
	Info.ViewerControl(VCTL_SETPOSITION, &vsp);
}

Int64 Viewer::SetFrame(Int64 pos, int left, ViewFrameOptions options)
{
	AssertCurrentViewer();
	ViewerSetPosition vsp;
	vsp.Flags = (DWORD)options;
	vsp.LeftPos = left;
	vsp.StartPos = pos;
	Info.ViewerControl(VCTL_SETPOSITION, &vsp);
	return vsp.StartPos;
}

void Viewer::Close()
{
	AssertCurrentViewer();
	Info.ViewerControl(VCTL_QUIT, 0);
}

void Viewer::Redraw()
{
	AssertCurrentViewer();
	Info.ViewerControl(VCTL_REDRAW, 0);
}

void Viewer::Select(Int64 symbolStart, int symbolCount)
{
	AssertCurrentViewer();
	if (symbolCount <= 0)
	{
		Info.ViewerControl(VCTL_SELECT, 0);
	}
	else
	{
		ViewerSelect vs;
		vs.BlockLen = symbolCount;
		vs.BlockStartPos = symbolStart;
		Info.ViewerControl(VCTL_SELECT, &vs);
	}
}

void Viewer::AssertClosed()
{
	if (IsOpened) throw gcnew InvalidOperationException("Viewer must not be open for this operation.");
}

bool Viewer::HexMode::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	if (vi.ViewerID < 0 || vi.ViewerID != _id)
		return false;
	else
		return vi.CurMode.Hex != 0;
}

void Viewer::HexMode::set(bool value)
{
	AssertCurrentViewer();
	ViewerSetMode vsm;
	vsm.Flags = vsm.Reserved = 0;
	vsm.Type = VSMT_HEX;
	vsm.Param.iParam = value;
	Info.ViewerControl(VCTL_SETMODE, &vsm);
}

bool Viewer::WrapMode::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	if (vi.ViewerID < 0 || vi.ViewerID != _id)
		return false;
	else
		return vi.CurMode.Wrap != 0;
}

void Viewer::WrapMode::set(bool value)
{
	AssertCurrentViewer();
	ViewerSetMode vsm;
	vsm.Flags = vsm.Reserved = 0;
	vsm.Type = VSMT_WRAP;
	vsm.Param.iParam = value;
	Info.ViewerControl(VCTL_SETMODE, &vsm);
}

bool Viewer::WordWrapMode::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	if (vi.ViewerID < 0 || vi.ViewerID != _id)
		return false;
	else
		return vi.CurMode.WordWrap != 0;
}

void Viewer::WordWrapMode::set(bool value)
{
	AssertCurrentViewer();
	ViewerSetMode vsm;
	vsm.Flags = vsm.Reserved = 0;
	vsm.Type = VSMT_WORDWRAP;
	vsm.Param.iParam = value;
	Info.ViewerControl(VCTL_SETMODE, &vsm);
}

int Viewer::CodePage::get()
{
	if (!IsOpened)
		return _CodePage;

	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	if (vi.ViewerID < 0 || vi.ViewerID != _id)
		return 0;

	return vi.CurMode.CodePage;
}

void Viewer::CodePage::set(int value)
{
	AssertClosed();

	_CodePage = value;
}

}
